authorization/invitation/service/invitation_service.go

```diff
@@ -162,3 +199,84 @@ func (s *invitationServiceImpl) Issue(ctx context.Context, issuingUserId uuid.UU
 
 	return nil
 }
+
+type invitationNotification struct {
+	invitation *invitationrepo.Invitation
+	roles      []string
+}
+
+// processTeamInviteNotifications sends an e-mail notification to a user.
+func (s *invitationServiceImpl) processTeamInviteNotifications(ctx context.Context, team *account.Identity, inviterName string,
+	notifications []invitationNotification, witURL string) {
+	teamName := team.IdentityResource.Name
+
+	spaceName, err := lookupSpaceName(ctx, witURL, *team.IdentityResource.ParentResourceID)
+	if err != nil {
+		// What to do here?? just log an error and die?
+	}
+
+	var messages []notification.Message
+
+	for _, n := range notifications {
+		acceptURL := fmt.Sprintf("%s/api/invitations/accept?code=%s", s.config.GetAuthServiceURL(), n.invitation.AcceptToken.String())
+
+		messages = append(messages, notification.NewTeamInvitationEmail(n.invitation.Identity.UserID.UUID.String(),
+			teamName,
+			inviterName,
+			spaceName,
+			acceptURL))
+	}
+
+	s.Services().NotificationService().SendMessagesAsync(ctx, messages)
+}
+
+// processSpaceInviteNotifications sends an e-mail notification to a user.
+func (s *invitationServiceImpl) processSpaceInviteNotifications(ctx context.Context, space *resource.Resource,
+	inviterName string, notifications []invitationNotification, witURL string) {
+
+	spaceName, err := lookupSpaceName(ctx, witURL, space.ResourceID)
+	if err != nil {
+		// What to do here?? just log an error and die?
+	}
+
+	var messages []notification.Message
+
+	for _, n := range notifications {
+		acceptURL := fmt.Sprintf("%s/api/invitations/accept?code=%s", s.config.GetAuthServiceURL(), n.invitation.AcceptToken.String())
+
+		messages = append(messages, notification.NewSpaceInvitationEmail(n.invitation.Identity.UserID.UUID.String(),
+			spaceName,
+			inviterName,
+			strings.Join(n.roles, ","),
+			acceptURL))
+	}
+
+	s.Services().NotificationService().SendMessagesAsync(ctx, messages)
+}
+
+// lookupSpaceName talks to the WIT service to retrieve a space record for the specified spaceID, then
+// returns the name of the space
+func lookupSpaceName(ctx context.Context, witURL string, spaceID string) (string, error) {
```

Comment:
Thinking out aloud : 

We should ideally make notification-service do this. I agree that might increase the number of calls to the WIT Space API from the notification-service - but that should be a problem which notification-service should solve ( by doing some local caching? ) :) 

The current approach has the advantage of avoiding unnecessary calls for sure. However, if we had to now pass the description, we would have to make code changes to two codebases :) 

For now, we can go ahead with this.

Reply:
In this case, it makes more sense to do the space lookup in auth once (refactoring the notification service to support caching seems to be a non-trivial exercise).  I think we should probably determine this on a case by case basis, weighing up the benefits of assembling the notification message package ourselves vs getting the notification service to do it.